### DISCUSSION TITLE
Cors error when running FastAPI inside docker

### DISCUSSION THREAD

**Discussion by abhinavkulkarni** [fastapi/fastapi#8532] [Questions] [answered]
Hi,

I am running into a CORS/network error when trying to reach one API from another when both of them are running inside respective docker containers.

I have a git repository demonstrating the issue. The repo has a README.md file that details the issue. 

```bash
git clone --single-branch --branch 2022-06-28 https://github.com/abhinavkulkarni/Stackoverflow-Issues.git
```

Essentially, the 2nd API has a redirect to the 1st one and is unable to reach it when using `HOST_IP` as their base IP.

How can I resolve this? What IP address (other than that of the host's) should I use so that I can reach one API from another?

Thanks!

**Comment by jonatasoli** [comment]
Hello,

You try use the same docker-compose in you APIs? There is a specific situation to you need use RedirectResponse? I think is more easy get return with request or httpx and send response.

Why you don't use fastapi cors?
```python
from fastapi.middleware.cors import CORSMiddleware
```

I try use your repository later.

**Comment by abhinavkulkarni** [comment]
Hey @jonatasoli,

Thanks for the reply. 

> You try use the same docker-compose in you APIs?

As you can see in the repository [here](https://github.com/abhinavkulkarni/Stackoverflow-Issues.git), I have two `docker-compose.yml` files for two APIs.

> Why you don't use fastapi cors?

I do use FastAPI CORS middleware with generous permissions.

> I think is more easy get return with request or httpx and send response.

Yes, using `httpx` does solve the problem, but I am still curious as to why `RedirectResponse` doesn't work.

Thanks!

**Comment by jonatasoli** [comment]
> As you can see in the repository [here](https://github.com/abhinavkulkarni/Stackoverflow-Issues.git), I have two `docker-compose.yml` files for two APIs.

Yes, but I should use only 1 file, and two api shared the same network.

> I do use FastAPI CORS middleware with generous permissions.
I see you code, you use starlet cors import don't fastapi import.

```python
from fastapi import FastAPI
import uvicorn
from starlette.middleware.cors import CORSMiddleware
from starlette.responses import RedirectResponse

app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_methods=["*"],
    allow_headers=["*"],
)
```

> Yes, using `httpx` does solve the problem, but I am still curious as to why `RedirectResponse` doesn't work.
I see, I see if fastapi have RedirectResponse in fastapi lib api.
edit:

I see in docs and you can use this example in fastapi docs.
https://fastapi.tiangolo.com/advanced/custom-response/?h=redirect#redirectresponse

In your repository you import starlette lib.

**Comment by JarroVGIT** [comment, marked as answer]
I am sorry, I've been wrecking my brain on what the problem is that you are experiencing. When using curl, I get a 307 redirect as well. This is expected, as you literally return a `RedirectResponse()` in your myapi-2/index endpoint. When going to the endpoint using a browser, I am redirected to myapi-1/index endpoint and see the message `"message":"Welcome to my API 1!",`. 

So, in short, I don't understand what the issue is. Is it that your curl command doesn't do anything else than showing you the 307 rather than the end response? Because you would need to add an `-L` command to curl for it to follow 307 redirects. See the below:

```
jarro@MBP-van-Jarro Stackoverflow-Issues % curl -v -L http://${HOST_IP}:8001/index

*   Trying 192.168.2.12:8001...
* Connected to 192.168.2.12 (192.168.2.12) port 8001 (#0)
> GET /index HTTP/1.1
> Host: 192.168.2.12:8001
> User-Agent: curl/7.79.1
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 307 Temporary Redirect
< date: Tue, 28 Jun 2022 19:59:16 GMT
< server: uvicorn
< content-length: 0
< location: http://192.168.2.12:8000/index
< 
* Connection #0 to host 192.168.2.12 left intact
* Issue another request to this URL: 'http://192.168.2.12:8000/index'
*   Trying 192.168.2.12:8000...
* Connected to 192.168.2.12 (192.168.2.12) port 8000 (#1)
> GET /index HTTP/1.1
> Host: 192.168.2.12:8000
> User-Agent: curl/7.79.1
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< date: Tue, 28 Jun 2022 19:59:16 GMT
< server: uvicorn
< content-length: 34
< content-type: application/json
< 
* Connection #1 to host 192.168.2.12 left intact
{"message":"Welcome to my API 1!"}%      
```

I don't see what this has to do with CORS or what the issue is at all to be honest.

**Comment by abhinavkulkarni** [comment]
@JarroVGIT: Thanks for your helpful response! The `-L` flag certainly solves the problem. Pardon my ignorance.

However, if I put another URL in `RedirectRespones`, I get a JSON response back without needing to specify a `-L` flag.

For e.g., at [line 25](https://github.com/abhinavkulkarni/Stackoverflow-Issues/blob/2022-06-28/myapi-2/main.py#L25),  if I put `https://reqres.in/api/users?page=1` in the redirect URL (instead of `f"{os.getenv('MYAPI1_URL')}/index"`), the redirect happens automatically, without `-L` flag. Why so?

Also, is there a way to do the same in Swagger? As you can see in the below image, I get an error. Swagger uses `curl` command that doesn't include `-L` flag.

![image](https://user-images.githubusercontent.com/1565547/176345166-6d3ec386-4b7b-4cd5-a20b-f2fc70658be2.png)

Hey @jonatasoli,

Thanks for your replies too. 

> Yes, but I should use only 1 file, and two api shared the same network.

@jonatasoli: The idea is to mimic the real-world deployment of these APIs. They may be deployed on the same host or on different hosts. In that case, each needs to have its own `docker-compose.yml` file. For e.g., one of them may be an auth API whereas the other may be a backend API for a web app that uses auth API for authentication and authorization. Each request to a protected resource may have a dependency injection to the auth API (hence the cooked up redirect scenario). Each would be scaled and deployed separately as per usage.

**Comment by jonatasoli** [comment]
I try in my computer and I noticed your url don't have http:// prefix, I change and found.

```python
@app.get("/index")
async def index2():
    _URL = os.getenv('MYAPI1_URL')
    return RedirectResponse(url=f"http://{_URL}/index")
```

**Comment by abhinavkulkarni** [comment]
Hey @jonatasoli,

If you run the commands in the `README.md` file as is, you will notice that `http://` prefix does exist in front of redirect URL. You can try putting a print statement just before the redirect call and print the URL.

You can see here that `http://` prefix is attached:

```bash
export MYAPI1_URL=http://${HOST_IP}:8000
export MYAPI2_URL=http://${HOST_IP}:8001
```

Hey @JarroVGIT: Any updates on this? This is really perplexing to me. It may very well be a Docker networking issue, but I am not sure how to go about debugging it.

**Comment by jonatasoli** [comment]
>If you run the commands in the README.md file as is, you will notice that http:// prefix does exist in front of redirect URL. You can try putting a print statement just before the redirect call and print the URL.
>

I changed to export variable and found too, please add [loguru lib](https://github.com/Delgan/loguru) in your requirements and copy the code bellow in myapi2/main.py and post the output this myapi2 here.

```python
import os

from fastapi import FastAPI
import uvicorn
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import RedirectResponse
from loguru import logger

app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_methods=["*"],
    allow_headers=["*"],
)


@app.get("/")
async def index1():
    return {"message": "Welcome to my API 2!"}


@app.get("/index")
async def index2():
    _URL = os.getenv('MYAPI1_URL')
    logger.debug(f"My url {_URL}")
    return RedirectResponse(url=f"{_URL}/index")

```

I only change imports and found for me in browser and in OpenAPI.

**Comment by JarroVGIT** [comment]
I cannot reproduce: 
<img width="1435" alt="image" src="https://user-images.githubusercontent.com/13659033/176755304-b79e4f64-ef7c-4246-bcec-fd3c7e71a9b6.png">

So, when following your instructions, I am able to:
- Use cURL correctly, and when using `-L` flag you instruct cURL to follow a redirect (a response that you expected to receive, btw).
- Use the browser, which successfully redirects me from myapi2/index to myapi1/index
- Use the automatic docs and execute myapi2/index from there (see above). 

My suggestion is to search for the solution in your local (docker) setup, it probably goes wrong somewhere in there. It at least is not a FastAPI issue.
